src/main/java/com/vlkan/pubsub/PubsubSubscriber.java

```diff
@@ -0,0 +1,222 @@
+import io.micrometer.core.instrument.MeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PubsubSubscriber {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PubsubSubscriber.class);
+
+    private final PubsubSubscriberConfig config;
+
+    private final PubsubClient client;
+
+    private final Scheduler scheduler;
+
+    private final PullResponseConsumer pullResponseConsumer;
+
+    private final PubsubPullRequest pullRequest;
+
+    private final StagedRateLimiter rateLimiter;
+
+    private final DistributionSummary permitWaitPeriodDistributions;
+
+    @FunctionalInterface
+    public interface PullResponseConsumer {
+
+        Mono<PubsubAckRequest> accept(PubsubPullResponse pullResponse);
+
+    }
+
+    private PubsubSubscriber(Builder builder) {
+        this.config = builder.config;
+        this.client = builder.client;
+        this.scheduler = builder.scheduler;
+        this.pullResponseConsumer = builder.pullResponseConsumer;
+        this.pullRequest = new PubsubPullRequest(true, config.getPullBufferSize());
+        String rateLimiterName = String.format("%s/%s", config.getProjectName(), config.getSubscriptionName());
+        this.rateLimiter = StagedRateLimiter.of(rateLimiterName, config.getRateLimiterSpec());
+        this.permitWaitPeriodDistributions = DistributionSummary
+                .builder(config.getMeterName())
+                .register(builder.meterRegistry);
+    }
+
+    public PubsubSubscriberConfig getConfig() {
+        return config;
+    }
+
+    public PubsubClient getClient() {
+        return client;
+    }
+
+    public Scheduler getScheduler() {
+        return scheduler;
+    }
+
+    public PullResponseConsumer getPullResponseConsumer() {
+        return pullResponseConsumer;
+    }
+
+    public Flux<PubsubSubscriberConsumption> start() {
+        return Flux
+                .just(new AtomicBoolean(false))
+                .flatMap(emptyBatchReceivedRef -> FluxHelpers
+                        .infiniteRange(BigDecimal.ZERO)
+                        .concatMap(requestIndex -> {
+                            boolean delayEnabled = emptyBatchReceivedRef.get();
+                            return delayEnabled
+                                    ? Mono.just(requestIndex).delayElement(config.getPullPeriod(), scheduler)
+                                    : Mono.just(requestIndex);
+                        })
+                        .flatMap(
+                                requestIndex -> pullAndConsumeAndAck(emptyBatchReceivedRef, requestIndex),
+                                config.getPullConcurrency()))
+                .subscribeOn(scheduler);
```

Comment:
why do you need to subscribe on specific scheduler?

Reply:
Right, this is redundant. I have already made `scheduler` optional and removed this line.